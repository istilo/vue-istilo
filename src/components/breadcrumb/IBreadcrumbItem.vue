<script setup lang="ts">
import { computed } from 'vue';

const props = defineProps<{
	/**
	 * Identifier for current page
	 */
	isActive?: boolean;

	/**
	 * Disabled the click event
	 */
	isDisabled?: boolean;

	/**
	 * Location where to redirect
	 */
	href: string;
}>();

const getClasses = computed(() => {
	return [props.isDisabled ? 'i-disabled' : ''];
});

const getTagName = computed(() => {
	return props.isActive ? 'span' : 'router-link';
});
</script>

<template>
	<li :class="getClasses">
		<component
			:is="getTagName"
			:to="!props.isActive && props.href ? 'props.href' : ''">
			<slot />
		</component>
	</li>
</template>
