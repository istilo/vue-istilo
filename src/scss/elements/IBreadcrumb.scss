@import '@/scss/helpers';
@import '@/scss/functions';
@import '@/scss/mixins';

// ============================================================
// Name:            Breadcrumb
// Description:     Component to create a breadcrumb navigation
//
// Component:       `#{$prefix}breadcrumb`
//
// States:          `#{$prefix}disabled`
//
// ============================================================

// Variables
// ============================================================

$breadcrumb-item-font-size: font-size(small) !default;
$breadcrumb-item-color: color(muted) !default;
$breadcrumb-item-hover-color: color(emphasis) !default;
$breadcrumb-item-hover-text-decoration: none !default;
$breadcrumb-item-active-color: color(primary) !default;

$breadcrumb-divider: '/' !default;
$breadcrumb-divider-margin-horizontal: margin(small) !default;
$breadcrumb-divider-color: border(color) !default;

/* ============================================================
   Component: Breadcrumb
   ============================================================ /

/*
 * 1. Allow items to wrap into the next line
 * 2. Reset list
 */

.#{$prefix}breadcrumb {
	display: flex;

	/* 1 */
	flex-wrap: wrap;

	/* 2 */
	padding: 0;
	margin: 0;
	list-style: none;
}

/*
 * Space is allocated solely based on content dimensions: 0 0 auto
 */

.#{$prefix}breadcrumb > * {
	flex: none;
}

/* Items
   ============================================================ */

.#{$prefix}breadcrumb > * > * {
	display: inline-block;
	font-size: $breadcrumb-item-font-size;
	color: $breadcrumb-item-color;
}

/* Hover + Focus */
.#{$prefix}breadcrumb > * > :hover,
.#{$prefix}breadcrumb > * > :focus {
	color: $breadcrumb-item-hover-color;
	text-decoration: $breadcrumb-item-hover-text-decoration;
}

/* Disabled */
.#{$prefix}breadcrumb > .#{$prefix}disabled > * {
	color: color(idle);
	cursor: not-allowed;
}

/* Active */
.#{$prefix}breadcrumb > :last-child > * {
	color: $breadcrumb-item-active-color;
}

@include respond-to-max(medium) {
	.#{$prefix}breadcrumb > :not(:last-child) > * {
		display: none;
	}
}

/*
 * Divider
 * `nth-child` makes it also work without JS if it's only one row
 */

.#{$prefix}breadcrumb > :nth-child(n + 2):not(.uis-first-column):before {
	content: $breadcrumb-divider;
	display: inline-block;
	margin: 0 $breadcrumb-divider-margin-horizontal;
	color: $breadcrumb-divider-color;
}

@include respond-to-max(medium) {
	.#{$prefix}breadcrumb > :nth-child(n + 2):not(.uis-first-column):before {
		display: none;
	}
}
