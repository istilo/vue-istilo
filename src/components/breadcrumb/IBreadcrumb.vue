<script setup lang="ts">
import { computed } from 'vue';
import IBreadcrumbItem from './IBreadcrumbItem.vue';

interface Items {
	href: string;
	isDisabled: boolean;
	label: string;
}

const props = defineProps<{
	/**
	 * An array of object to display link
	 *
	 * @example { href : string, isDisabled : boolean, label : string }
	 */
	items: Items[];
}>();

const isLastItem = computed(() => {
	return props.items.length - 1;
});
</script>

<template>
	<ul class="i-breadcrumb">
		<i-breadcrumb-item
			v-for="(item, index) in items"
			:key="index"
			:isActive="isLastItem === index"
			:href="item.href"
			:isDisabled="item.isDisabled">
			{{ item.label }}
		</i-breadcrumb-item>
	</ul>
</template>
