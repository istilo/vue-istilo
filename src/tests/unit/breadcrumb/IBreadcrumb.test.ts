import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { shallowMount, VueWrapper, enableAutoUnmount } from '@vue/test-utils';

import IBreadcrumb from '@/components/breadcrumb/IBreadcrumb.vue';
import IBreadcrumbItem from '@/components/breadcrumb/IBreadcrumbItem.vue';

enableAutoUnmount(afterEach);

describe('Istilo Breadcrumb', () => {
	let wrapper: VueWrapper<any>;

	beforeEach(() => {
		wrapper = shallowMount(IBreadcrumb, {
			propsData: {
				items: [
					{
						href: '/home',
						isDisabled: false,
						label: 'Home',
					},
					{
						href: '/profile',
						isDisabled: false,
						label: 'Profile',
					},
					{
						href: '/profile/john-doe',
						isDisabled: false,
						label: 'John Doe',
					},
				],
			},
		});

		expect(wrapper.exists()).toBe(true);
	});
	afterEach(() => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('Should render the default class and tagname', () => {
		expect(wrapper.element.tagName).toBe('UL');
		expect(wrapper.classes()).toContain('i-breadcrumb');
	});

	it('Should render the children', () => {
		expect(wrapper.findAllComponents(IBreadcrumbItem)).toHaveLength(3);
	});
});
