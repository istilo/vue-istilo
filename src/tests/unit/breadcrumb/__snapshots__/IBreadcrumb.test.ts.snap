// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Istilo Breadcrumb > Should render the children 1`] = `
"<ul class=\\"i-breadcrumb\\">
  <i-breadcrumb-item-stub href=\\"/home\\" isactive=\\"false\\" isdisabled=\\"false\\"></i-breadcrumb-item-stub>
  <i-breadcrumb-item-stub href=\\"/profile\\" isactive=\\"false\\" isdisabled=\\"false\\"></i-breadcrumb-item-stub>
  <i-breadcrumb-item-stub href=\\"/profile/john-doe\\" isactive=\\"true\\" isdisabled=\\"false\\"></i-breadcrumb-item-stub>
</ul>"
`;

exports[`Istilo Breadcrumb > Should render the default class and tagname 1`] = `
"<ul class=\\"i-breadcrumb\\">
  <i-breadcrumb-item-stub href=\\"/home\\" isactive=\\"false\\" isdisabled=\\"false\\"></i-breadcrumb-item-stub>
  <i-breadcrumb-item-stub href=\\"/profile\\" isactive=\\"false\\" isdisabled=\\"false\\"></i-breadcrumb-item-stub>
  <i-breadcrumb-item-stub href=\\"/profile/john-doe\\" isactive=\\"true\\" isdisabled=\\"false\\"></i-breadcrumb-item-stub>
</ul>"
`;
